share/eds/p2p/server.go

```diff
@@ -0,0 +1,163 @@
+package p2p
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/protocol"
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/eds"
+	p2p_pb "github.com/celestiaorg/celestia-node/share/eds/p2p/pb"
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+)
+
+const (
+	// writeDeadline sets timeout for sending messages to the stream
+	writeDeadline = time.Second * 5
+	// readDeadline sets timeout for reading messages from the stream
+	readDeadline = time.Minute
+)
+
+// Server is responsible for serving EDSs for blocksync over the ShrEx/EDS protocol.
+// This client is run by bridge nodes and full nodes. For more information, see ADR #13
+type Server struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	host       host.Host
+	protocolID protocol.ID
+
+	store *eds.Store
+}
+
+// NewServer creates a new ShrEx/EDS server.
+func NewServer(host host.Host, store *eds.Store, protocolSuffix string) *Server {
+	return &Server{
+		host:       host,
+		protocolID: protocolID(protocolSuffix),
+		store:      store,
+	}
+}
+
+func (s *Server) Start(context.Context) error {
+	s.ctx, s.cancel = context.WithCancel(context.Background())
+	s.host.SetStreamHandler(s.protocolID, s.handleStream)
+	return nil
+}
+
+func (s *Server) Stop(context.Context) error {
+	defer s.cancel()
+	s.host.RemoveStreamHandler(s.protocolID)
+	return nil
+}
+
+func (s *Server) handleStream(stream network.Stream) {
+	log.Debug("server: handling eds request")
+
+	// read request from stream to get the dataHash for store lookup
+	req, err := s.readRequest(stream)
+	if err != nil {
+		log.Errorw("server: reading request from stream", "err", err)
+		stream.Reset() //nolint:errcheck
+		return
+	}
+
+	var edsReader io.ReadCloser
+	status := p2p_pb.Status_OK
+	err = share.DataHash(req.Hash).Validate()
+	if err != nil {
+		status = p2p_pb.Status_INVALID
```

Comment:
Status Invalid does not mean to be used in practice. It's a 0-value code and is a PB convention to prevent uninitialized enums from being sent over the wire. 

If the data hash is invalid(it's size is less than 32) it means we have a malicious client, and we should just reset the stream without caring about responding to anything. 

Will add this to the document

Reply:
done (resets stream now)